rpc: serve tenants and node using single cert and port

tbg opened the pull request:
This PR simplifies the multi-tenancy deployment by serving tenants on
the same port as intra-cluster traffic. This is desirable since it keeps
the deployment for a multi-tenant CockroachDB KV layer close to that of
a single-tenant cluster.

We originally served on distinct ports due to the complication of
setting up a TLS-enabled config on a single port when it was necessary
to route incoming connections to either of two certificates. The first
problem to solve here is that this has to be done based solely on
[ClientHelloInfo], which presently is not populated with any information
to tell incoming tenant connections apart from other KV nodes'; I
believe the canonical way to solve this would have been to start
populating the `SupportedProtos` field, which we could have used to
determine the set of certs to offer to the client.
The second problem is essentially the first, but for client certs: for
an incoming connection, we need to determine whether to treat the
client as a tenant or internal KV connection. This we have already
solved: as of #52602, tenant client certs have `OrganizationalUnit`
set to `Tenants`, which is visible to the auth handlers.

A recent change (#52516) mandating that the tenant server would
always use the internal node certificates made the work to solve problem
1) unnecessary - if there is only one cert, no switching is necessary.
We can simply merge the two servers into one (doing the same for the
authorization handlers), which is what this PR achieves.

The door to in the future allowing any combination of separate
port/separate certs remains open since we can start using the
`SupportedProtos` field if necessary. In the medium term, we do not
anticipate this becoming necessary.

[ClientHelloInfo]: https://golang.org/pkg/crypto/tls/#ClientHelloInfo

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52508)
<!-- Reviewable:end -->

tbg commented:
Ignore the first commit - that's https://github.com/cockroachdb/cockroach/pull/52478

darinpp commented on pkg/rpc/auth_tenant.go:
```diff
@@ -79,7 +80,11 @@ func (a tenantAuth) tenantFromCert(ctx context.Context) (roachpb.TenantID, error
 
 	commonName := tlsInfo.State.PeerCertificates[0].Subject.CommonName
-	tenID, err := strconv.ParseUint(commonName, 10, 64)
+	const prefix = "tenant_"
```
I will change this so we don't rely on prefix (that would make hard to distinguish tenant from an SQL user named tenant_33). The type of the CN will be in OU. I.e. `OU=Tenants CN=33`

darinpp reviewed (commented):
For 1 - The split port functionality is not needed right now. So I'm fine with ripping it out now and adding it when needed. 
For 2 - I also think this is simpler to be always on.

Even though we don't have this case and don't want to do it - It also isn't a problem to have tenant certs minted by different CAs. The KV nodes can add the tenant CA (ca-tenant-client.crt will be present and contain the tenant CA's cert in this case) to pool and authenticate clients using certs issued by either CA. See how this is done for the split client CA in `security.newServerTLSConfig`

nvanbenschoten commented:
> If the answers to 1) and 2) are "no" and "no", we can phase out
SplitListenTenant and melt kvAuth and tenantAuth into a single
authenticator that avoids the awkward delegation logic.

I agree with @darinpp. "no" and "no" seems like the right answer right now. And that means that we can push the delegation of authorization policies into a single auth that delegates based on cert common name.

tbg committed:
rpc: serve tenants and node using single cert and port

This PR simplifies the multi-tenancy deployment by serving tenants on
the same port as intra-cluster traffic. This is desirable since it keeps
the deployment for a multi-tenant CockroachDB KV layer close to that of
a single-tenant cluster.

We originally served on distinct ports due to the complication of
setting up a TLS-enabled config on a single port when it was necessary
to route incoming connections to either of two certificates. The first
problem to solve here is that this has to be done based solely on
[ClientHelloInfo], which presently is not populated with any information
to tell incoming tenant connections apart from other KV nodes'; I
believe the canonical way to solve this would have been to start
populating the `SupportedProtos` field, which we could have used to
determine the set of certs to offer to the client.
The second problem is essentially the first, but for client certs: for
an incoming connection, we need to determine whether to treat the
client as a tenant or internal KV connection. This we have already
solved: as of #52602, tenant client certs have `OrganizationalUnit`
set to `Tenants`, which is visible to the auth handlers.

A recent change (#52516) mandating that the tenant server would
always use the internal node certificates made the work to solve problem
1) unnecessary - if there is only one cert, no switching is necessary.
We can simply merge the two servers into one (doing the same for the
authorization handlers), which is what this PR achieves.

The door to in the future allowing any combination of separate
port/separate certs remains open since we can start using the
`SupportedProtos` field if necessary. In the medium term, we do not
anticipate this becoming necessary.

[ClientHelloInfo]: https://golang.org/pkg/crypto/tls/#ClientHelloInfo

Release note: None

tbg committed:
security: remove vestigial tenant server cert helpers

These became essentially unused in #52516.

Release note: None

tbg commented:
Ok, I did most of the work. It's not quite ready to merge yet but should turn mostly green. @darinpp you could give this a first review pass or take it for a spin. I hope to polish this tomorrow.

darinpp commented on pkg/rpc/context.go:
```diff
@@ -114,29 +114,25 @@ func spanInclusionFuncForClient(
+			opts.interceptor = f
+		} else {
+			opts.interceptor = func(fullMethod string) error {
+				if err := opts.interceptor(fullMethod); err != nil {
```
Seems to me that this will lead to infinite recursion as written. The new function will capture just the `opts`. The `interceptor` will be the function itself after the assignment.

darinpp reviewed (approved):
Looks good.

tbg commented:
Thanks for the review, Darin. I went in and cleaned up the authentication code, I think it came out pretty well. All changes are in a separate commit that I will squash before merging, but definitely should get another look.

I also noticed something else that might possibly be a problem. We are going to have "node certs" for the KV layer and each SQL pod. If they are issued by the same CA, won't they work interchangeably? In effect, we will have certificates on the SQL pods that give full access to the KV layer. I don't know what we were planning to do here, but essentially the only certs issued by "the" CA that should be available to the SQL pod are the tenant certs, nothing else.

For example, the latest version of `demo.sh` does have this problem: `certs-sql/node.*` gives root access to KV when used as a client cert.

```bash
#!/bin/bash

set -euxo pipefail

killall -9 cockroach || true

rm -rf certs-{sql,kv,user} cockroach-data
mkdir -p certs-{sql,kv,user}
sql="certs-sql"
kv="certs-kv"
cakey="${kv}/ca.key"

#### KV certs

./cockroach cert --certs-dir="${kv}" --ca-key="${cakey}" create-ca
./cockroach cert --certs-dir="${kv}" --ca-key="${cakey}" create-node 127.0.0.1 ::1 localhost *.local


#### SQL tenant certs

ln -s "../${kv}/ca.crt" "${sql}"
# Certs that SQL server uses to contact KV as the tenant.
./cockroach mt cert --certs-dir="${sql}" --ca-key="${cakey}" create-tenant-client 10
# Certs that SQL server uses to serve incoming connections (from the proxy).
#
# TODO(darinpp): in practice these must NOT use the same CA, for they look like node certs and are accessible
# to the user should the pod be compromised! These must not be allowed to connect to the KV layer. We need to
# use a separate CA here.
./cockroach cert --certs-dir="${sql}" --ca-key="${cakey}" create-node 127.0.0.1 ::1 localhost *.local

#### Root client cert

# NB: this gives full access to the KV layer, so end users must never get a hold of it. (Unless we're using a
# separate ClientCA that the KV layer does not know about).
./cockroach cert --certs-dir="${sql}" --ca-key="${cakey}" create-client root
mv "${sql}"/client.root.* certs-user
ln -s "../${sql}/ca.crt" certs-user

#### Start KV and create tenant

./cockroach start-single-node --certs-dir certs-kv --host 127.0.0.1 --background
./cockroach sql --certs-dir certs-user --host 127.0.0.1 -e 'select crdb_internal.create_tenant(10);'

#### Start SQL and create 'bob' using root client certs

./cockroach mt start-sql --certs-dir certs-sql --tenant-id 10 --kv-addrs 127.0.0.1:26257 --sql-addr 127.0.0.1:46257 &
sleep 1

./cockroach sql --url 'postgres://root@127.0.0.1:46257?sslmode=verify-ca&sslkey=certs-user/client.root.key&sslcert=certs-user/client.root.crt&sslrootcert=certs-user/ca.crt' -e "CREATE USER bob WITH PASSWORD 'foo';"


#### Start SQL proxy

./cockroach mt start-proxy --listen-addr 127.0.0.1:56257 --target-addr 127.0.0.1:46257 --listen-cert certs-sql/node.crt --listen-key certs-sql/node.key &
sleep 1

#### Connect as 'bob' through proxy, once using db, once using options.

./cockroach sql --url 'postgres://bob:foo@127.0.0.1:56257/prancing-pony.defaultdb?sslmode=verify-ca&sslrootcert=certs-user/ca.crt' -e "SELECT 'this worked!';"

./cockroach sql --url 'postgres://bob:foo@127.0.0.1:56257?options=--cluster=prancing-pony&sslmode=verify-ca&sslrootcert=certs-user/ca.crt' -e "SELECT 'that worked!';"

echo "Everything worked! Here are all the CA certs involved:"
find certs-* -type f -name 'ca*'
```

darinpp commented:
Yes, this is a valid concern. The immediate plan is to restrict the interchangeability by restricting the DNS name in the cert. The current subject alternative name that we put on the kv node certs has a long list of DNS names, including `DNS:*.cockroachdb`. for the MVP - the SQL pods work with the same cert that is used by the KV nodes, so I had to put them in the same namespace as the kv nodes - cockroachdb. Now that we will mint individual certs for each SQL pod, this isn't required (and is desirable to change) anymore. A cert for the tenant cluster `dim-dog` will have subject alternative name of `DNS:dim-dog-0.tenants` (assuming that we move all tenant pods into the tenants namespace). So none of the tenant certs are interchangeable from one tenant to another and they are also not interchangeable with the cert used by the KV nodes.
Ideally - it would make a lot of sense to add OU=Nodes to the KV-only node certs and to add OU=SQLNode to the SQL-only node certs (and to do checks on the server side). This would strengthen even more the use of certificates for the intended purpose only.

tbg commented:
@nvanbenschoten doesn't have to be today, but mind giving this a look as well? In particular the changes to `kvAuth`.

nvb reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> All in all, I'm happy we got to delete all of this. Thanks for pushing this through.

> Reviewed 31 of 45 files at r3, 15 of 15 files at r5, 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52508#-:-MEx2f7cAcGrRLSQnMS1:b-nsfw62)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

tbg commented:
bors r=nvanbenschoten
